misc/bazel/pkg.bzl

```diff
@@ -1,4 +1,316 @@
+"""
+Wrappers and helpers around `rules_pkg` to build codeql packs.
+"""
+
+load("@rules_pkg//pkg:install.bzl", "pkg_install")
+load("@rules_pkg//pkg:mappings.bzl", "pkg_attributes", "pkg_filegroup", "pkg_files", _strip_prefix = "strip_prefix")
+load("@rules_pkg//pkg:pkg.bzl", "pkg_zip")
 load("@rules_pkg//pkg:providers.bzl", "PackageFilegroupInfo", "PackageFilesInfo")
+load("@rules_python//python:defs.bzl", "py_binary")
+
+def _make_internal(name):
+    def internal(suffix = "internal"):
+        return "%s-%s" % (name, suffix)
+
+    return internal
+
+_PLAT_DETECTION_ATTRS = {
+    "_windows": attr.label(default = "@platforms//os:windows"),
+    "_macos": attr.label(default = "@platforms//os:macos"),
+}
+
+_PLAT_PLACEHOLDER = "{CODEQL_PLATFORM}"
+
+def _process_path(path, plat):
+    if _PLAT_PLACEHOLDER in path:
+        path = path.replace(_PLAT_PLACEHOLDER, plat)
+        return ("arch", path)
+    return ("generic", path)
+
+def _detect_plat(ctx):
+    if ctx.target_platform_has_constraint(ctx.attr._windows[platform_common.ConstraintValueInfo]):
+        return "windows64"
+    elif ctx.target_platform_has_constraint(ctx.attr._macos[platform_common.ConstraintValueInfo]):
+        return "osx64"
+    else:
+        return "linux64"
+
+def codeql_pkg_files(
+        *,
+        name,
+        srcs = None,
+        exes = None,
+        visibility = None,
+        **kwargs):
+    """ Wrapper around `pkg_files` adding a distinction between `srcs` and `exes`, where the
+    latter will get executable permissions.
+    """
+
+    internal = _make_internal(name)
+    if "attributes" in kwargs:
+        fail("do not use attributes with codeql_pkg_* rules. Use `exes` to mark executable files.")
+    internal_srcs = []
+    if srcs and exes:
+        pkg_files(
+            name = internal("srcs"),
+            srcs = srcs,
+            visibility = ["//visibility:private"],
+            **kwargs
+        )
+        pkg_files(
+            name = internal("exes"),
+            srcs = exes,
+            visibility = ["//visibility:private"],
+            attributes = pkg_attributes(mode = "755"),
+            **kwargs
+        )
+        pkg_filegroup(
+            name = name,
+            srcs = [internal("srcs"), internal("exes")],
+            visibility = visibility,
+        )
+    else:
+        pkg_files(
```

Comment:
I think it's the case that `pkg_files` can be used wherever `pkg_filegroup` can as far as `rules_pkg` is concerned. In particular it can go into `srcs` of  `pkg_filegroup`, `pkg_install` or all archive creators (`pkg_zip` and co). Unless I'm missing a simple example of the contrary...

Reply:
(I do consume a `PackageFilegroupInfo` in `_extract_pkg_filegroup_impl`, but that's used in a macro that explicitly instantiates a `pkg_filegroup`, so I don't have to follow the same rule there)